jxl/src/headers/permutation.rs

```diff
@@ -0,0 +1,336 @@
+            prev_val = val;
+        }
+
+        // Initialize the full permutation vector with skipped elements intact
+        let mut permutation: Vec<u32> = Vec::new();
+        permutation.try_reserve((size - skip) as usize)?;
+        permutation.extend(0..size);
+
+        // Decode the Lehmer code into the slice starting at `skip`
+        let permuted_slice = decode_lehmer_code(&lehmer, &permutation[skip as usize..])?;
+
+        // Replace the target slice in `permutation`
+        permutation.splice(skip as usize.., permuted_slice);
+
+        // Ensure the permutation has the correct size
+        assert_eq!(permutation.len(), size as usize);
+
+        Ok(Self(permutation))
+    }
+}
+
+// Decodes the Lehmer code in `code` and returns the permuted slice.
+fn decode_lehmer_code(code: &[u32], permutation_slice: &[u32]) -> Result<Vec<u32>> {
+    let n = permutation_slice.len();
+    if n == 0 {
+        return Err(Error::InvalidPermutationLehmerCode {
+            size: 0,
+            idx: 0,
+            lehmer: 0,
+        });
+    }
+
+    let mut permuted = permutation_slice.to_vec();
+    let log2n = (n as u32).ceil_log2();
+    let padded_n = 1 << log2n;
+
+    // Allocate temp array inside the function
+    let mut temp = vec![0u32; padded_n];
+
+    // Initialize temp array
+    for (i, temp_item) in temp.iter_mut().enumerate().take(padded_n) {
+        let i1 = (i + 1) as u32;
+        *temp_item = value_of_lowest_1_bit(i1);
+    }
+
+    for (i, permuted_item) in permuted.iter_mut().enumerate().take(n) {
+        let code_i = *code.get(i).unwrap_or(&0);
+
+        // Adjust the maximum allowed value for code_i
+        if code_i as usize > n - i - 1 {
+            return Err(Error::InvalidPermutationLehmerCode {
+                size: n as u32,
+                idx: i as u32,
+                lehmer: code_i,
+            });
+        }
+
+        let mut rank = code_i + 1;
+
+        // Extract i-th unused element via implicit order-statistics tree.
+        let mut bit = padded_n;
+        let mut next = 0usize;
+        while bit != 0 {
+            let cand = next + bit;
+            if cand == 0 || cand > padded_n {
+                return Err(Error::InvalidPermutationLehmerCode {
+                    size: n as u32,
+                    idx: i as u32,
+                    lehmer: code_i,
+                });
+            }
+            bit >>= 1;
+            if temp[cand - 1] < rank {
+                next = cand;
+                rank -= temp[cand - 1];
+            }
+        }
+
+        *permuted_item = permutation_slice[next];
+
+        next += 1;
+        while next <= padded_n {
+            temp[next - 1] -= 1;
+            next += value_of_lowest_1_bit(next as u32) as usize;
+        }
+    }
+
+    Ok(permuted)
+}
+
+// Decodes the Lehmer code in `code` and returns the permuted vector.
+#[allow(dead_code)]
+fn decode_lehmer_code_naive(code: &[u32], permutation_slice: &[u32]) -> Result<Vec<u32>> {
```

Comment:
It would be good to move this into `mod test` block.

Reply:
good idea.
Instead I now have added `#[cfg(test)]`